bam_markdup.c

```diff
@@ -467,6 +627,200 @@ static int add_duplicate(khash_t(duplicates) *d_hash, bam1_t *dupe) {
+    }
+
+    if (strncmp(original, duplicate, oxpos - 1) == 0) {
+        // the initial parts match, look at the numbers
+        long ox, oy, dx, dy, xdiff, ydiff;
+        char *end;
+
+        ox = strtol(original + oxpos, &end, 10);
+
+        if ((original + oxpos) == end) {
+            fprintf(stderr, "[markdup] warning: can not decipher X coordinate in %s .\n", original);
+            return ret;
+        }
+
+        dx = strtol(duplicate + dxpos, &end, 10);
+
+        if ((duplicate + dxpos) == end) {
+            fprintf(stderr, "[markdup] warning: can not decipher X coordinate in %s.\n", duplicate);
+            return ret;
+        }
+
+        if (ox > dx) {
+            xdiff = ox - dx;
+        } else {
+            xdiff = dx - ox;
+        }
+
+        if (xdiff <= max_dist) {
+            // still might be optical
+
+            oy = strtol(original + oypos, &end, 10);
+
+            if ((original + oypos) == end) {
+                fprintf(stderr, "[markdup] warning: can not decipher Y coordinate in %s.\n", original);
+                return ret;
+            }
+
+            dy = strtol(duplicate + dypos, &end, 10);
+
+            if ((duplicate + dypos) == end) {
+                fprintf(stderr, "[markdup] warning: can not decipher Y coordinate in %s.\n", duplicate);
+                return ret;
+            }
+
+            if (oy > dy) {
+                ydiff = oy - dy;
+            } else {
+                ydiff = dy - oy;
+            }
+
+            if (ydiff <= max_dist) ret = 1;
+        }
+    }
+
+    return ret;
+}
+
+
+static int mark_duplicates(md_param_t *param, khash_t(duplicates) *dup_hash, bam1_t *ori, bam1_t *dup, int *optical) {
+    char dup_type = 0;
+
+    dup->core.flag |= BAM_FDUP;
+
+    if (param->tag) {
+        if (bam_aux_append(dup, "do", 'Z', strlen(bam_get_qname(ori)) + 1, (uint8_t*)bam_get_qname(ori))) {
+            fprintf(stderr, "[markdup] error: unable to append 'do' tag.\n");
+            return -1;
+        }
+    }
+
+    if (param->opt_dist) { // mark optical duplicates
+        if (optical_duplicate(ori, dup, param->opt_dist)) {
+            bam_aux_append(dup, "dt", 'Z', 3, (const uint8_t *)"SQ");
+            dup_type = 'O';
+            (*optical)++;
+        } else {
+            // not an optical duplicate
+            bam_aux_append(dup, "dt", 'Z', 3, (const uint8_t *)"LB");
+        }
+    }
+
+    if (param->supp) {
+        if (bam_aux_get(dup, "SA") || (dup->core.flag & BAM_FMUNMAP) || bam_aux_get(dup, "XA")) {
+            char *original = NULL;
+
+            if (param->tag) {
+                original = bam_get_qname(ori);
+            }
+
+            if (add_duplicate(dup_hash, dup, original, dup_type))
+                return -1;
+        }
+    }
+
+    return 0;
+}
+
+
+/* Modified Lander-Waterman equation */
+static inline double lw_equation(double x, double c, double n) {
+    return c / x - 1 + exp(-n / x);
+}
+
+
+/* estimate the library size */
```

Comment:
As this is essentially the Picard code reimplemented in C, we should put in a comment to say where it came from, and also add them to the copyright at the top of the file.  Something like:

```
Estimate library size derived from Picard DuplicationMetrics.java
  Copyright (c) 2009,2018 The Broad Institute.  MIT license.
```

Luckily they use the same license as us.

Reply:
True, though I'm not sure how useful the library estimate is.